File: C++/Sorting/sorting algorithms/SelectionSort.cpp
//
//  SelectionSort.cpp
//  Sorting
//
//  Created by Paul Kraft on 08.04.16.
//  Copyright © 2016 pauljohanneskraft. All rights reserved.
//

#include "SelectionSort.hpp"

SelectionSort::SelectionSort() : Sort("SelectionSort\t\t") {}

void SelectionSort::sort(Comparable *array[], int length) {
    sort(array, 0, length -1);
}

void SelectionSort::sort(Comparable *array[], int left, int right) {
    for(int i = left; i <=right; i++) {
        int min = i;
        for(int j = i+1; j <= right; j++) {
            if(*array[j] < array[min]) { min = j; }
        }
        Comparable *tmp = array[i];
        array[i] = array[min];
        array[min] = tmp;
    }
}

File: Swift/Sorting/sorting algorithms/SelectionSort.swift
//
//  SelectionSort.swift
//  Sort
//
//  Created by Paul Kraft on 21.12.15.
//  Copyright © 2015 Paul Kraft. All rights reserved.
//

extension SortableCollection {
    public mutating func selectionSort(by order: (Element, Element) throws -> Bool) rethrows {
        // ...
        try self.selectionSort(in: self.indices, by: order)
    }
    
    public mutating func selectionSort(in range: CountableRange<Int>, by order: (Element, Element) throws -> Bool) rethrows {
        // ...
        // an array with 1 element doesn't need to be sorted
        guard range.count > 1 else { return }
        
        for l in range {
            var min = l
            for i in l+1..<range.endIndex {
                if try order(self[i], self[min]) { min = i }
            }
            if l != min { swap(&self[l], &self[min]) }
        }
    }
}

extension SortableCollection where Element : Comparable {
    public mutating func selectionSort() {
        self.selectionSort(by: { $0 < $1 })
    }
}
